README
#
# What this extension does
# ========================
#
# - A special page "Graphical Categories Browser" is added.
# - A graph is shown on each category page showing the relationships to other
#   categories.
# - Images are file cached. Cache control to match html and image output is
#   supplied.
#
#
# Installation
# ============
#
# - Install Graphviz - Graph Visualization Software available at
#   http://www.graphviz.org/
# - Download extension and copy to to directory extensions.
# - Correct $xyDotPath and $xyCategoriesCache to your needs.
# - Create directory images/xyGraphvizCache
# - Add the following line to LocalSettings.php
#   require_once("$IP/extensions/xyCategoryBrowser/xyCategoryBrowser.php");
# - Developer documentation can be generated in directory doc using Doxygen
#   with file doxyfile.
#
# Customization
# =============
#
# To change the design you may override the entries in the following array:
#
# $xyCategoryGraphStyle = array(
#   "COLOR_NODE"          => "#EEEEEE", // color of category nodes
#   "COLOR_NODE_ERROR"    => "#FF0000", // color for internal error
#   "COLOR_NODE_REDIRECT" => "#FFCCCC", // color of redirected category nodes
#   "COLOR_NODE_MISSING"  => "#FFFFCC", // color of missing category nodes
#   "COLOR_LINK_REDIRECT" => "#FF0000", // color of redirect links
#   "HEIGHT"              => "1920",    // height in pixels (96th of an inch)
#   "WIDTH"               => "768"      // width in pixels (96th of an inch)
# );







ChangeLog
Moved from SourceForge to github to ease installation in Dantewiki


Verison 1.9, 2022-12-25
	Replace new article() by WikiPage::factory()
	Replace Title::newFromRedirect() by Content::getRedirectTarget()
Version 1.8, 2018-03-28
	Replace wfMsg() by wfMessage()->text()
Version 1.7, 2014-12-08
	Correct version information
	Check if 'MEDIAWIKI' is defined.
Version 1.6, 2013-04-08
	Refactoring (file layout, use class based special pages)
	Use UTF8 instead of HTML entities
Version 1.5, 2012-11-02
	Added DIV around category browser
	Removed date and link to GraphViz
Version 1.4, 2012-10-19
	Corrected display of non existent categories
	Color codes moved to constants
Version 1.3, 2012-04-06
	License changed
Version 1.2, 2012-04-05
	Code adjusted for current MediaWiki release


CategoryGraph.php
<?php

class CategoryGraph {

// that is the main function where we enter
public static function onCategoryPageView( &$cat ) { 
  $title = $cat->getTitle();
  $dbKey = $title->getDBkey();

  $dot = self::doQuery($title);
  self::doDot($dbKey, $dot);
  self::showImg($dbKey);
 }



static  private function getSQLCategories( $title = null ) {  // Constructs SQL statement to select categories related to title.
    global $wgOut;
 
    $id   = $title->getArticleID();
    $text = $title->getDBkey();
 
    $NScat = NS_CATEGORY;
    $dbr = wfGetDB(DB_REPLICA);
    // Use the following for MediaWiki 1.9:
    $text = $dbr->addQuotes($text);
    //$text = "'".wfStrencode($text)."'"; 
 
    $categorylinks = $dbr->tableName('categorylinks');
    $page          = $dbr->tableName('page');
    $sql =
      "SELECT\n".
      "    page_title AS cat,\n".
      "    page_is_redirect AS redirect,\n".
      "    0                AS missing\n".
      "  FROM $page as a\n".
      "  left JOIN $categorylinks as b\n".
      "  ON a.page_id=b.cl_from\n".
      "  left join $categorylinks as c\n".
      "  ON a.page_title=c.cl_to\n".
      "  WHERE\n".
      "    page_namespace = {$NScat} AND\n".
      "  (  c.cl_from    = $id OR\n".
      "     a.page_id    = $id OR\n".
      "     b.cl_to      = {$text} )\n".
      "UNION\n".
      "SELECT\n".
      "    cl_to as cat,\n".
      "    0 AS redirect,\n".
      "    1 AS missing\n".
      "  FROM $categorylinks\n".
      "  LEFT JOIN $page\n".
      "  ON page_title=cl_to\n".
      "  WHERE\n".
      "    page_id IS NULL";


  // cl_to:    name of the category to which the link is pointing to
  // cl_fom:   page id of the page which contains the category link

  // missing is true for all categories, which are employed in category links but do not have a category page describing it


  // $wgOut->addHTML("<pre>$sql</pre>");
  return $sql;
}

static private function getSQLCategoryLinks( $title ) { // Constructs SQL statement to select links between categories.
  global $wgOut;
 
  $id    = $title->getArticleID();
  $text  = $title->getDBkey();
  $NScat = NS_CATEGORY;
  $dbr   = wfGetDB(DB_REPLICA);

  // Use the following for MediaWiki 1.9:
  $text = $dbr->addQuotes($text);
  // $text = "'".wfStrencode($text)."'"; 
 
  $categorylinks = $dbr->tableName('categorylinks');
  $page          = $dbr->tableName('page');
  $sql =
      "SELECT\n".
      "    page_title AS cat_from, \n".
      "    cl_to as cat_to\n".
      "  FROM $page\n".
      "  INNER JOIN $categorylinks\n".
      "  ON page_id=cl_from\n".
      "  WHERE\n".
      "    ( page_id=$id  OR\n".
      "    cl_to=$text ) AND\n".
      "    page_namespace={$NScat}";

   // $wgOut->addHTML("<pre>$sql</pre>");
  return $sql;
}
  

  /**
   * @brief Embeds category graph into page.
   *
   * @param title page title
   */
static  function doQuery($title = null) {
    global $wgOut;

$xyCategoryGraphStyle = array(
   "COLOR_NODE"          => "#00EEEE", // color of category nodes
   "COLOR_NODE_ERROR"    => "#FF0000", // color for internal error
   "COLOR_NODE_REDIRECT" => "#FFCCCC", // color of redirected category nodes
   "COLOR_NODE_MISSING"  => "#FF0000", // color of missing category nodes
   "COLOR_LINK_REDIRECT" => "#FF0000", // color of redirect links
   "HEIGHT"              => "1920",    // height in pixels (96th of an inch)
   "WIDTH"               => "768"      // width in pixels (96th of an inch)
 );

  $colorNode         = $xyCategoryGraphStyle["COLOR_NODE"];
  $colorNodeError    = $xyCategoryGraphStyle["COLOR_NODE_ERROR"];
  $colorNodeRedirect = $xyCategoryGraphStyle["COLOR_NODE_REDIRECT"];
  $colorNodeMissing  = $xyCategoryGraphStyle["COLOR_NODE_MISSING"];
  $colorLinkRedirect = $xyCategoryGraphStyle["COLOR_LINK_REDIRECT"];
  $height            = $xyCategoryGraphStyle["HEIGHT"]/96;
  $width             = $xyCategoryGraphStyle["WIDTH"]/96;
 
  $redirections= Array();
  $nodes=Array();
  
  # Start digraph and set defaults
  $dot = "digraph a {\nsize=\"{$width},{$height}\";\nrankdir=LR;\n".
         "node [height=0 style=\"filled\", shape=\"box\", ".
         "font=\"Helvetica-Bold\", fontsize=\"10\", color=\"#00000\"];\n";

  $dbr = wfGetDB( DB_REPLICA );
  $sql= self::getSQLCategories($title);
  $res = $dbr->query($sql, __METHOD__);          // Run the query; returns IResultWrapper or false     // TODO: missing error handling in case of false

  danteLog ("DanteCategoryBrowser", "first query: number of rows: " . $res->numRows ($res) . "\n");
  for ( $i = 0; $obj = $res->fetchRow($res); $i++ ) {
    danteLog ("DanteCategoryBrowser", "first query, object $i: " . print_r ($obj, true) . "\n");

    $l_title = Title::makeTitle(NS_CATEGORY, $obj["cat"]);
 
    $color = $colorNode;
    if($obj["redirect"]==1)  $color = $colorNodeRedirect;
    if($obj["missing"]== 1)  $color = $colorNodeMissing;

    $nodes[$obj["cat"]] = array(
      'color' => $color,
      'url'   => $l_title->getFullURL(),
      'peri'  => 1,
      'label' => str_replace( '_', ' ', $obj["cat"] )
    );

     danteLog ("DanteCategoryBrowser", "first query: node: " . print_r ($nodes[$obj["cat"]] , true)."\n");


    if ($title && $obj["cat"] == $title->getDBkey()) { $nodes[$obj["cat"]]['peri']=2; }

    if ($obj["redirect"]) {
      $article = WikiPage::factory($l_title);
      if ($article) {
        $text = $article->getContent();
        $rt = $text->getRedirectTarget();
        if ($rt) {
          if (NS_CATEGORY == $rt->getNamespace()) {
            $redirections[$l_title->getDBkey()] = $rt->getDBkey();
            if (!$nodes[$rt->getDBkey()]){
              $nodes[$rt->getDBkey()] = array(
                'color' => $colorNode,
                'url'   => $rt->getFullURL(),
                'peri'  => 1
                );
                }
              }
            }
          }
    }
  }  // end for loop over result set

  $sql= self::getSQLCategoryLinks($title);
  $res = $dbr->query($sql, __METHOD__);          // Run the query; returns IResultWrapper or false     // TODO: missing error handling in case of false

  danteLog ("DanteCategoryBrowser", "SECOND QUERY number of rows: " . $res->numRows ($res) . "\n");

  for ( $i = 0; $obj = $res->fetchRow( $res ); $i++ ) {
    danteLog ("DanteCategoryBrowser", "second query object: " . print_r ($obj, true) . "\n");

    $cat_from = Title::makeName(NS_CATEGORY, $obj["cat_from"]);
    $cat_to   = Title::makeName(NS_CATEGORY, $obj["cat_to"]);

    # If destination node has not been read highlight the error.  // TODO ?????
      if (@!$nodes[$obj["cat_to"]]){
        $rt = Title::makeTitle(NS_CATEGORY, $obj["cat_to"]);
        $nodes[$rt->getDBkey()] = array(
          'color' => $colorNodeError,
          'url'   => $rt->getFullURL(),
          'peri'  => ($title && $rt->getDBkey() == $title->getDBkey())? 2 : 1,
          'label' => str_replace( '_', ' ', $rt->getDBkey())
          );
        }
      else {
        danteLog ("DanteCategoryBrowser", "not highlioghted : " . print_r ($obj, true) . "\n");
      }


    if (
  //          !$redirections[$obj["cat_from"]] 
! isset($redirections[$obj['cat_from']])
|| 
 $redirections[$obj["cat_from"]] != $obj["cat_to"]) {
      $dot .= "\"".$obj['cat_to']."\" -> \"".$obj['cat_from']."\" [dir=back];\n";
    }
    else {danteLog ("DanteCategoryBrowser", "not redirected : " . print_r ($obj, true) . "\n"); }


  }  // end for loop second query



  // Create redirection links
  foreach( $redirections as $cat_from => $cat_to) {
    $dot .= "\"$cat_to\" -> \"$cat_from\" [color=\"".$colorLinkRedirect."\", dir=back];\n";
  }

  foreach( $nodes as $l_DbKey=>$properties ) {   
    $l_title = Title::makeTitle(NS_CATEGORY, $l_DbKey);
    $dot .= "\"$l_DbKey\" [URL=\"{$properties['url']}\",".
        "peripheries={$properties['peri']},label=\"{$properties['label']}\",".
        "fillcolor=\"{$properties['color']}\"];\n";
  }
 
  $dot .= "}\n";
  //$wgOut->addHTML("<pre>${dot}</pre>"); // debug  this is the graphviz dot code
  return $dot;
}


static private function doDot( $title, $dot ) {  // Save dot file and generate png and map file
  global $wgOut, $danteDotPath;
  global $IP;
 
  $md5 = md5($title);
  $docRoot = "$IP/danteCatBrowserCache/";

  $fileDot = "$docRoot$md5.dot";
  $fileMap = "$docRoot$md5.map";
  $filePng = "$docRoot$md5.png";

  file_put_contents($fileDot, $dot);

  $cmd = "$danteDotPath -Tpng -o$filePng <$fileDot";

  // $wgOut->addHTML("$danteDotPath -Tpng -o$filePng <$fileDot");  // debug
  $result = shell_exec($cmd);

  danteLog ("DanteCategoryBrowser", "calling $cmd gave $result \n");

  // $wgOut->addHTML("$danteDotPath -Tcmap -o$fileMap <$fileDot");  // debug
  $map = shell_exec("$danteDotPath -Tcmap -o$fileMap <$fileDot");
}


public static function onRegistration () { 
  global $IP;
  if ( !file_exists ($IP."/danteCatBrowserCache") ) {$retVal = mkdir ($IP."/danteCatBrowserCache", 0755);}
}


static function showImg( $title ) {  // output the image; title to generate md5 for filename
  global $wgOut, $wgScriptPath, $wgServer;
  global $IP;
  $docRoot = "$IP/danteCatBrowserCache/";
  $md5 = md5($title);
  $fileMap = "$docRoot$md5.map";
  if (true || file_exists($fileMap)) {
    $map = file_get_contents($fileMap);

    $mapStr = htmlspecialchars ($map);
    //$wgOut->addHTML("<pre>$mapStr</pre>");  // debug
 
    $URLpng =  $wgServer.$wgScriptPath."/danteCatBrowserCache/$md5.png";
   //$wgOut->addHTML("<pre>URL: $URLpng</pre>");  // debug


    $wgOut->addHTML("<DIV style='border:1px solid green;' id='xyCategoryBrowser'><IMG src=\"$URLpng\" usemap=\"#map1\" alt=\"$title\"><MAP name=\"map1\">$map</MAP></DIV>");
    return true;
    }
}

}
?>


xyCategoryGraph.php
<?php
/**
* @author  Heinrich Schuchardt
* @version 1.7 - 2014-12-08
* @file xyCategoryGraph.php
* @ingroup Extension
* @license http://www.gnu.org/copyleft/gpl.html GNU General Public License 3 or later
*
* Copyright (C) 2013, Heinrich Schuchardt
*
* Changes:
* 1.7 Update license information
* 1.6 Created
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser Public License for more details.
*
* You should have received a copy of the GNU Lesser Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/

/**
 * @brief Creates graph for category page.
 */
require_once 'xyAbstractCategoryGraph.php';
class xyCategoryGraph extends xyAbstractCategoryGraph{

  /**
   * @brief Creates new instance.
   */
  public function __construct() {
    parent::__construct();
  }

  /**
   * @brief Constructs SQL statement to select categories related to title.
   *
   * @param $title title
   * @return SQL statement
   */
  function getSQLCategories( $title = null ) {
    global $wgOut;
 
    $id   = $title->getArticleID();
    $text = $title->getDBkey();
 
    $NScat = NS_CATEGORY;
    $dbr =& wfGetDB(DB_REPLICA);
    // Use the following for MediaWiki 1.9:
    $text = $dbr->addQuotes($text);
    //$text = "'".wfStrencode($text)."'"; 
 
    $categorylinks = $dbr->tableName('categorylinks');
    $page          = $dbr->tableName('page');
    $sql =
      "SELECT\n".
      "    page_title AS cat,\n".
      "    page_is_redirect AS redirect,\n".
      "    0                AS missing\n".
      "  FROM $page as a\n".
      "  left JOIN $categorylinks as b\n".
      "  ON a.page_id=b.cl_from\n".
      "  left join $categorylinks as c\n".
      "  ON a.page_title=c.cl_to\n".
      "  WHERE\n".
      "    page_namespace = {$NScat} AND\n".
      "  (  c.cl_from    = $id OR\n".
      "     a.page_id    = $id OR\n".
      "     b.cl_to      = {$text} )\n".
      "UNION\n".
      "SELECT\n".
      "    cl_to as cat,\n".
      "    0 AS redirect,\n".
      "    1 AS missing\n".
      "  FROM $categorylinks\n".
      "  LEFT JOIN $page\n".
      "  ON page_title=cl_to\n".
      "  WHERE\n".
      "    page_id IS NULL";

    if ($this->debug) $wgOut->addHTML("<"."pre>$sql<"."/pre>");
    return $sql;
    }

  /**
   * @brief Constructs SQL statement to select links between categories.
   *
   * @param $title title
   * @return SQL statement
   */ 
  function getSQLCategoryLinks( $title ) {
    global $wgOut;
 
    $id   = $title->getArticleID();
    $text = $title->getDBkey();
 
    $NScat = NS_CATEGORY;
    $dbr =& wfGetDB(DB_REPLICA);
    // Use the following for MediaWiki 1.9:
    $text = $dbr->addQuotes($text);
    // $text = "'".wfStrencode($text)."'"; 
 
    $categorylinks = $dbr->tableName('categorylinks');
    $page          = $dbr->tableName('page');
    $sql =
      "SELECT\n".
      "    page_title AS cat_from, \n".
      "    cl_to as cat_to\n".
      "  FROM $page\n".
      "  INNER JOIN $categorylinks\n".
      "  ON page_id=cl_from\n".
      "  WHERE\n".
      "    ( page_id=$id  OR\n".
      "    cl_to=$text ) AND\n".
      "    page_namespace={$NScat}";
    if ($this->debug) $wgOut->addHTML("<"."pre>$sql<"."/pre>");
    return $sql;
    }
  }
?>


xyCategoriesPage.php
<?php
/**
* @author  xypron
* @version 1.7 - 2014-12-08
* @file xyCategoriesPage.php
* @ingroup Extension
* @license http://www.gnu.org/copyleft/gpl.html GNU General Public License 3 or later
*
* Copyright (C) 2013, Heinrich Schuchardt
*
* Changes:
* 1.7 Update license information
*     Check MEDIAWIKI is defined
* 1.6 Created
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser Public License for more details.
*
* You should have received a copy of the GNU Lesser Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/

if(!defined('MEDIAWIKI'))
  die("This is not a valid entry point to MediaWiki.\n");

require_once 'xyAbstractCategoryGraph.php';

/**
 * @brief Creates content for special page with graph of all categories.
 */
class xyCategoriesPage extends xyAbstractCategoryGraph{

  /**
   * @brief Constructs SQL statement to select categories related to title.
   *
   * @param $title title
   * @return SQL statement
   */
  function getSQLCategories( $title = null ) {
    global $wgOut;

    $NScat = NS_CATEGORY;
    $dbr =& wfGetDB(DB_REPLICA);
    $categorylinks = $dbr->tableName('categorylinks');
    $page          = $dbr->tableName('page');
    $sql =
      "SELECT\n".
      "    page_title AS cat,\n".
      "    page_is_redirect AS redirect,\n".
      "    0 AS missing\n".
      "  FROM $page\n".
      "  WHERE\n".
      "    page_namespace={$NScat}\n".
      "UNION\n".
      "SELECT\n".
      "    cl_to as cat,\n".
      "    0 AS redirect,\n".
      "    1 AS missing\n".
      "  FROM $categorylinks\n".
      "  LEFT JOIN $page\n".
      "  ON page_title=cl_to\n".
      "  WHERE\n".
      "    page_id IS NULL";
    if ($this->debug) $wgOut->addHTML("<"."pre>$sql<"."/pre>");
    return $sql;
  }

  /**
   * @brief Constructs SQL statement to select links between categories.
   *
   * @param $title title
   * @return SQL statement
   */
  function getSQLCategoryLinks( $title = null ) {
    global $wgOut;

    $NScat = NS_CATEGORY;
    $dbr =& wfGetDB(DB_REPLICA);
    $categorylinks = $dbr->tableName('categorylinks');
    $page          = $dbr->tableName('page');
    $sql =
      "SELECT\n".
      "    page_title AS cat_from, \n".
      "    cl_to as cat_to,\n".
      "    page_is_redirect AS redirect\n".
      "  FROM $page\n".
      "  INNER JOIN $categorylinks\n".
      "  ON page_id=cl_from\n".
      "  WHERE\n".
      "    page_namespace={$NScat}";
    if ($this->debug) $wgOut->addHTML("<"."pre>$sql<"."/pre>");

    return $sql;
  }
}

?>


SpecialExtensionSpecialPage.php
<?php



class SpecialPageClass extends SpecialPage {
  public function __construct() {
    parent::__construct( 'DanteCategoryBrowser_SpecialPage' );   // only one parameter; second parameter would be a restriction on rights
    danteLog ("DanteCategoryBrowser", "constructed\n");
  }



	/**
	 * Show the page to the user
	 *
	 * @param string $sub The subpage string argument (if any).
	 */
	public function execute( $sub ) {
  danteLog ("DanteCategoryBrowser", "in execute\n");  // looks like this is never called

		$request = $this->getRequest();
		$output = $this->getOutput();
		$this->setHeaders();

		# Get request data from, e.g.
		$param = $request->getText( 'param' );

		# Do stuff
		# ...
		$wikitext = 'Hello world!';
		$output->addWikiTextAsInterface( $wikitext );



		$out = $this->getOutput();

//		$out->setPageTitle( $this->msg( 'special-extensionSpecialPage-title' ) );
//		$out->addHelpLink( 'How to become a MediaWiki hacker' );
//		$out->addWikiMsg( 'special-extensionSpecialPage-intro' );
	}

// define the group under which the page will show up in "Special:SpecialPages
// we here use "dante". This means that in the i18n file, here or somewhere else, we have to declare a "specialpages-group-dante"
protected function getGroupName() {
  return 'dante';
  }


}







xyCategoryBrowser.php
<?php
/**
* @author  Heinrich Schuchardt
* @version 1.7 - 2014-12-08
* @file xyCategoryBrowser.php
* @ingroup Extension
* @license http://www.gnu.org/copyleft/gpl.html GNU General Public License 3 or later
*
* Copyright (C) 2012, Heinrich Schuchardt
*
* Changes:
* 1.7 Update license and credits information
* 1.6 Refactored
* 1.5 Added DIV around category browser
*     Removed date and link to GraphViz
* 1.4 Corrected display of non existent categories
*     Color codes moved to constants
* 1.3 License changed
* 1.2 Created
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser Public License for more details.
*
* You should have received a copy of the GNU Lesser Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/

/**
 * @brief Path to the GraphViz dot executable
 */
// Linux
$xyDotPath = "/usr/bin/dot";
// Windows
//$xyDotPath = "\"C:\\Program Files\\ATT\\Graphviz\\bin\\dot.exe\"";

/**
 * @brief Age after which the graph is newly rendered in seconds. 
 */
$xyCategoriesMaxAge = 36;
/**
 * @brief Path for the file cache relative to this script.
 */
$xyCategoriesCache = "/../../images/xyGraphvizCache/";
/**
 * @brief Style of category graphs.
 */
$xyCategoryGraphStyle = array(
  "COLOR_NODE"          => "#EEEEEE", // color of category nodes
  "COLOR_NODE_ERROR"    => "#FF0000", // color for internal error
  "COLOR_NODE_REDIRECT" => "#FFCCCC", // color of redirected category nodes
  "COLOR_NODE_MISSING"  => "#FFFFCC", // color of missing category nodes
  "COLOR_LINK_REDIRECT" => "#FF0000", // color of redirect links
  "HEIGHT"              => "1920",    // height in pixels (96th of an inch)
  "WIDTH"               => "768"      // width in pixels (96th of an inch)
);
 
if(!defined('MEDIAWIKI')){
  require_once 'xyCategoryGraph.php';
  // Serve the PNG image
  $cap = new xyCategoryGraph();
  if ($cap->serveFile()) die();
  header("HTTP/1.1 404 Not Found");
  die("<H1>404 Not Found </H1>");
  }
 
//install special page
$wgExtensionFunctions[] = 'xyCategoryHook::hookSetup';

$wgExtensionCredits['other'][] = array(
    'name' => 'Graphical Category Browser',
    'version' => '1.7',
    'author' => 'Heinrich Schuchardt',
    'url' => 'http://www.mediawiki.org/wiki/Extension:Graphical_Category_Browser',
    'description' => 'A graph is added on top of each category page showing the relationships to other categories.',
    'license-name' => 'GPL-3.0+'
    );  

$wgExtensionCredits['specialpage'][] = array(
    'name' => 'Graphical Category Browser',
    'version' => '1.7',
    'author' => 'Heinrich Schuchardt',
    'url' => 'http://www.mediawiki.org/wiki/Extension:Graphical_Category_Browser',
    'description' => 'A special page "Graphical Categories Browser" is added.',
    'license-name' => 'GPL-3.0+'
    );  

// Internationalization file
$wgExtensionMessagesFiles['Xygraphicalcategorybrowser'] = __DIR__ .
    '/xyCategoryBrowser.i18n.php';
// Classes to be loaded automatically
$wgAutoloadClasses['xyCategoriesSpecialPage'] = __DIR__ .
    '/xyCategoriesSpecialPage.php'; 
$wgSpecialPages['Xygraphicalcategorybrowser'] = 'xyCategoriesSpecialPage';
$wgSpecialPageGroups['Xygraphicalcategorybrowser'] = 'pages';

/**
 * @brief Supplies hook function to add graph to category pages.
 */
class xyCategoryHook {
 
// DEPRECATE
/*
  public static function hookSetup() {
    global $IP, $wgMessageCache, $wgHooks;
 
    $wgHooks['CategoryPageView'][] = 'xyCategoryHook::hook';
  }
*/

  /**
   * @brief Hook function adds graph to category pages.
   *
   * @param $cat category
   */ 
  public static function hook($cat) {
    require_once 'xyCategoryGraph.php';
    global $wgOut, $xyCategoriesMaxAge;
    $wgOut->lowerCdnMaxage( $xyCategoriesMaxAge );
    $title = $cat->getTitle();
    $dbKey = $title->getDBkey();
    $cap = new xyCategoryGraph();
    $age = $cap->cacheAge($dbKey);
    if (!$age || $age > $xyCategoriesMaxAge ) {
      $cap->doQuery($title);
      $cap->doDot($dbKey);
      };
    $cap->showImg($dbKey);
    return true;
    }
  }
?>
